docs/metrics-api-design.md

```diff
@@ -0,0 +1,519 @@
+};
+
+class IntValueObserver: public AsynchronousInstrument{
+public:
+    /*
+    * Add adds the value to the counter's sum. The labels should contain
+    * the keys and values to be associated with this value.  Counters only     * accept positive valued updates.
+    * 
+    * @param value the numerical representation of the metric being captured
+    * @param labels the set of labels, as key-value pairs
+    */
+    void observe(int value, std::string_view &labels) override;
+private:
+    // Callback function which takes a pointer to an Asynchronous instrument (this) 
+    // type which is stored in an observer result type and returns nothing.  
+    // The function calls the instrument's observe method.
+    void (*callback)(IntObserverResult);
+}
+```
+
+```
+// The above Counter and BoundCounter are examples of 1 metric instrument.  
+// The remaining 5 will also be implemented in a similar fashion. With Int and 
+// Double for each metric instrument
+
+class IntUpDownCounter: public SynchronousInstrument;
+class DoubleUpDownCounter: public SynchronousInstrument;
+
+class IntValueRecorder: public SynchronousInstrument;
+class DoubleValueRecorder: public SynchronousInstrument;
+
+class IntSumObserver: public AsynchronousInstrument;
+class DoubleSumObserver: public AsynchronousInstrument;
+
+class IntUpDownSumObserver: public AsynchronousInstrument;
+class DoubleUpDownSumObserver: public AsynchronousInstrument;
+
+class IntValueObserver: public AsynchronousInstrument;
+class DoubleValueObserver: public AynchronousInstrument;
+
+
+
+class BoundIntUpDownCounter: public BoundSynchronousInstrument;
+class BoundDoubleUpDownCounter: public BoundSynchronousInstrument;
+
+class BoundIntValueRecorder: public BoundSynchronousInstrument;
+class BoundDoubleValueRecorder: public BoundSynchronousInstrument;
+```
+
+### **Metric Class Design Considerations**:
+
+OpenTelemetry requires several types of metric instruments with very similar core usage, but slightly different tracking schemes.  As such, a base Metric class defines the necessary functions for each instrument leaving the implementation for the specific instrument type.  Each instrument then inherits from this base class making the necessary modifications.  In order to facilitate efficient aggregation of labeled data, a complementary BoundInstrument class is included which attaches the same set of labels to each capture.  Knowing that all data in an instrument has the same labels enhances the efficiency of any post-collection calculations as there is no need for filtering or separation.  In the above code examples, a Counter instrument is shown but all 6 mandated by the specification will be supported.
+
+A base BoundInstrument class also serves as the foundation for more specific bound instruments.  It also facilitates the practice of reference counting which can determine when an instrument is unused and can improve memory optimization as inactive instruments can be removed for performance.
```

Comment:
I don't understand the note about reference counting. It seems we return metric instruments as `unique_ptr`s, so by definition there can only be one owner. I don't see how this relates to a `BoundInstrument` base class.

Reply:
After discussing this more, we've found that we may need to return metric instruments as shared pointers. This is because, in our SDK design, we will be placing pointers to the metric instruments in a container stored in the Meter class as well as returning the pointer to the instrument. It could also reduce the complexity by not having to do manual reference counting. 

Do you think it would be a good choice to use a shared ptr here?